### ISSUE TITLE
`networking.domain` is ignored

### ISSUE THREAD

**Issue by blitz** [NixOS/nixpkgs#87633]
**Describe the bug**

`networking.domain` is ignored and results in e.g. `hostname --fqdn` to not report the correct domain. This results in problems down the road, because tools like Geary, the Gnome email client, cannot send emails because it fails to send a FQDN in its EHLO STMP request.

**To Reproduce**
1. Set the following in NixOS configuration and rebuild the system:

```
  networking.hostName = "foo";
  networking.domain = "bar";
```

2. `/etc/hostname` contains the hostname without domain (unlike Fedora)
3. `hostname --fqdn` returns the hostname without the domain

**Expected behavior**

`hostname --fqdn` shows a fully-qualified domain name. 

**Notify maintainers**

@Infinisil @abbradar @lheckemann 

**Metadata**

 - system: `"x86_64-linux"`
 - host os: `Linux 5.6.11, NixOS, 20.03.1822.5adf2a6c116 (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.4`
 - channels(root): `"nixos-20.03.1822.5adf2a6c116, nixos-hardware, unstable-20.09pre223023.fce7562cf46"`
 - channels(julian): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
 -  network-interfaces
```

**Comment by emilazy** [comment]
FWIW, I have `networking.domain` set and it seems to work fine for me:

```
emily@renko ~> hostname --fqdn
renko.emily.moe
emily@renko ~> cat /etc/hostname
renko
```

So there might be some additional configuration details making this break.

**Comment by emilazy** [comment]
Looking at the implementation:

```nix
{
  # ...

  config = {
    # ...

    # Set the host and domain names in the activation script.  Don't
    # clear it if it's not configured in the NixOS configuration,
    # since it may have been set by dhcpcd in the meantime.
    system.activationScripts.hostname =
      optionalString (cfg.hostName != "") ''
        hostname "${cfg.hostName}"
      '';
    system.activationScripts.domain =
      optionalString (cfg.domain != null) ''
        domainname "${cfg.domain}"
      '';
  };

  # ...
}
```

It would be useful to know what network manager you're using (dhcpcd, connman, networkd, NetworkManager, etc.), how your network interfaces are configured, and whether `/etc/resolv.conf` contains a `domain ...` line.

**Comment by primeos** [comment]
@blitz yes, unfortunately the current behaviour is far from ideal :o

There's already #10183 (and probably other PRs/issues) so I'll close this issue to avoid splitting the discussion, but thanks for the report.
My attempt to fix this was #76542. It worked for me and I didn't see any big issues with that approach, but others seem to prefer using `nss-myhostname` instead (but I didn't have time to have another look at this).
